### ISSUE TITLE
[RFC]: Memory Operations as a Learned Sequential Decision Policy — Context-Adaptive Memory Management via MDP

### ISSUE THREAD

**Issue by JasperHG90** [JasperHG90/memex#263]
## Problem Statement

Memex's memory management uses fixed heuristics: always retrieve top-K by similarity, consolidate when a threshold is reached, forget when memory exceeds budget. These heuristics are static — they don't change based on task context, agent state, or task progression stage. Recent research (MemCon, arXiv:2607.13591, Jiang et al., July 2026) demonstrates that this static approach is a core bottleneck because **optimal memory behaviour is fundamentally context-dependent**:

1. **Early task stages** benefit from *minimal retrieval* — memory is sparse, and retrieving irrelevant content adds noise.
2. **Recurring goal types** benefit from *plan reuse* — retrieving a past successful plan is better than generic nearest-neighbor lookup.
3. **Stuck agents** benefit from *re-retrieval with alternative queries* — rephrasing the query and retrieving again surfaces different memory that might unblock progress.
4. **Long task streams** require *consolidation and pruning* — without it, the memory store becomes bloated and retrieval quality degrades.

No fixed heuristic handles all four cases. A system that always retrieves top-K misses the plan-reuse opportunity. A system that always consolidates prevents the stuck agent from re-retrieving raw memory. The optimal strategy depends on the current context.

Memex's proposed self-optimizing memory (#248) uses feedback signals to tune memory parameters (similarity threshold, consolidation frequency). But #248 tunes *parameters* — it doesn't choose *operations*. The question isn't "what similarity threshold to use?" but "should I retrieve, re-retrieve, consolidate, or forget right now?" This is a sequential decision problem, not a parameter tuning problem.

MemCon's key insight: model memory operations as a **Markov Decision Process (MDP)** and learn a policy that adaptively chooses operations. The state is the current task context + memory state. The actions are memory operations (retrieve, re-retrieve with alternative query, inject distilled plan, consolidate, forget). The reward is task success/failure. The policy learns which operations to take in which states, and it does so with a lightweight tabular contextual bandit — no LLM calls, no pretraining, convergence within tens of tasks.

## Proposed Solution

Implement **memory operations as a learned sequential decision policy** in Memex:

1. **MDP formulation**: Model memory management as an MDP:
   - **State**: `(task_context, memory_state, agent_state)` where:
     - `task_context`: Current task type, goal, progress estimate
     - `memory_state`: Memory size, density, last-consolidation time, retrieval count
     - `agent_state`: Number of steps taken, whether stuck (repeated failures), error count
   - **Actions**: Memory operations:
     - `retrieve_topk`: Standard top-K similarity retrieval (current default)
     - `retrieve_rephrase`: Re-retrieve with an alternative query (rephrased or reformulated)
     - `inject_plan`: Retrieve and inject a past successful plan for similar goals
     - `consolidate`: Trigger consolidation of recent memory
     - `forget`: Trigger eviction of low-importance memory
     - `no_op`: Don't access memory this step
   - **Reward**: Task success (+1) / failure (-1), with intermediate progress signals from #260 (Reward Densification) if available.

2. **Contextual bandit policy**: Learn the policy using a tabular contextual bandit with UCB (Upper Confidence Bound) exploration:
   - Discretise the state space into contexts (task type × memory size bucket × agent state bucket)
   - For each context, maintain action values (estimated reward for each action)
   - Select actions using UCB: choose the action with the highest upper confidence bound, balancing exploitation (best known action) with exploration (try less-known actions)
   - Update action values after task completion based on the reward
   - Convergence within tens of tasks (per MemCon's results)

3. **Backend-agnostic wrapper**: Following MemCon's design, the policy wraps any existing memory implementation:
   - The memory backend (vector store, graph store, etc.) provides the primitive operations
   - The policy decides *which* operation to call and *with what parameters*
   - No changes to the memory backend are needed — the policy is a middleware layer
   - This makes it compatible with existing Memex memory infrastructure

4. **Stuck-agent detection**: Implement a stuck detector that triggers re-retrieval:
   - Track consecutive failures (tool calls that error, repeated identical queries, loops in agent behaviour)
   - When the agent is stuck, the policy can choose `retrieve_rephrase` — reformulate the query and retrieve again
   - The rephrasing can use an LLM (expensive) or simple query expansion (cheap: add/remove terms, use entity names from the context)

5. **Plan injection**: For recurring goal types, the policy can choose `inject_plan`:
   - Identify past successful tasks with similar goals
   - Extract the plan (sequence of actions that led to success)
   - Inject the plan into the current context as a suggested approach
   - This is different from retrieval — it's not retrieving *facts* but retrieving *strategies*

### Design sketch

```
memex_core/memory/
├── mdp_policy.py            # New: MDP formulation + contextual bandit
├── stuck_detector.py        # New: detect when agent is stuck
├── plan_injection.py        # New: retrieve and inject past successful plans
└── query_rephraser.py       # New: alternative query generation for re-retrieval

memex_core/services/
├── search.py                # Existing: wrapped by policy
├── consolidation.py         # Existing: wrapped by policy
└── outcomes.py              # Existing: provide reward signal to policy
```

### Key design decisions

- **Policy over operations, not parameters**: The policy chooses *which operation* to perform, not just *what parameters* to use. This is a higher-level decision than #248's parameter tuning. #248 might tune the similarity threshold; this policy decides whether to retrieve at all, or to consolidate instead.
- **Tabular bandit, not neural network**: The policy is a simple tabular contextual bandit, not a neural network. This follows MemCon's finding that a lightweight tabular approach converges within tens of tasks with no pretraining. No GPU, no training pipeline, no model storage.
- **UCB exploration**: The UCB formula ensures that the policy explores actions it hasn't tried much in a given context, preventing it from getting stuck in a suboptimal strategy. This is critical for adapting to new task types.
- **Backend-agnostic**: The policy wraps the memory backend, not replaces it. This means it can be added to Memex without changing the existing memory infrastructure.
- **No additional LLM calls**: The policy itself makes no LLM calls. It uses the task outcome (success/failure) as the reward signal. LLM calls only happen if the `retrieve_rephrase` action uses LLM-based query reformulation (optional — simple query expansion also works).

## Alternatives Considered

1. **Fixed heuristics (current approach)**: Rejected — MemCon shows that context-dependent memory behaviour consistently outperforms fixed heuristics by up to 15.2 points in task success across 6 benchmarks.

2. **Extend #248 (Self-Optimizing Memory) to choose operations**: #248 tunes parameters within a single operation type (e.g., adjusting similarity threshold for retrieval). Extending it to choose between operations (retrieve vs. consolidate vs. forget) would require modelling the sequential decision problem — which is exactly what this RFC does. The MDP formulation is the natural framework for sequential decisions, not parameter tuning.

3. **RL with neural network policy**: Use a deep RL approach with a neural network policy. Rejected — MemCon shows that a tabular bandit suffices and converges in tens of tasks. A neural network would require more data, more compute, and more complexity without clear benefit.

4. **LLM-based memory management**: Ask an LLM to decide which memory operation to perform. Rejected — too expensive for every memory decision, and MemCon shows that a cheap tabular policy works better.

5. **Rule-based context detection**: Write rules that map contexts to operations (e.g., "if agent is stuck, re-retrieve"). Rejected — rules are brittle and don't adapt to new task types. The learned policy adapts automatically.

## Additional Context

**Source paper**: Memory as a Controlled Process: Learned Adaptive Memory Management for LLM Agents (arXiv:2607.13591, Jiang et al., July 2026)

**Key results from the paper**: MemCon consistently outperforms multiple memory baselines by up to 15.2 points in task success while reducing token consumption by 5–20% across 6 benchmarks, 3 agent frameworks, and 3 LLM backbones. The tabular contextual bandit converges within tens of tasks with no pretraining and no additional LLM calls.

**Related existing issues**:
- #248 (Self-Optimizing Memory) — parameter tuning; this RFC adds operation selection (a higher-level decision)
- #220 (Adaptive Memory Lifecycle) — lifecycle management; this RFC adaptively triggers lifecycle operations
- #260 (Reward Densification) — provides per-operation reward signals; this RFC uses them for policy learning
- #254 (Temporal Credit Assignment) — attributes outcomes to memory units; this RFC attributes outcomes to memory *decisions*
- #257 (Load-Bearing Memory Detection) — load-bearing detection protects critical memory; this policy decides when to forget
- #261 (Distributed Retrieval Signal Propagation) — graph-based retrieval; this policy decides when to use graph propagation vs. standard retrieval

**Related codebase modules**:
- `memex_core/services/search.py` — retrieval; wrapped by policy
- `memex_core/services/consolidation.py` — consolidation; wrapped by policy
- `memex_core/services/outcomes.py` — task outcomes; provide reward signal
- `memex_core/memory/confidence.py` — confidence; policy decisions feed into confidence
- `memex_core/memory/engine.py` — memory engine; policy is middleware layer

**Research note**: MemCon's key contribution is framing memory management as a **sequential decision problem** (MDP) rather than a parameter tuning problem. The insight that "optimal memory behaviour is fundamentally context-dependent" and that "a tabular bandit converges within tens of tasks" makes this approach practical for Memex — it doesn't need GPUs, training pipelines, or LLM calls. The backend-agnostic wrapper design means it can be added as a middleware layer without changing existing memory infrastructure.

**Comment by JasperHG90** [comment]
Closing for now as part of the RFC backlog triage. A learned operation policy needs task volume and a clean reward signal to converge — a single-user deployment produces neither, and the reward plumbing (#254, optionally #260) is not built. One small idea is worth stealing without the MDP framing: stuck-agent detection triggering a re-retrieval with a rephrased query is a cheap standalone heuristic. Reopen after the learning-signal foundations exist.
